compiler/rustc_expand/src/expand.rs

```diff
@@ -387,6 +393,18 @@ pub struct MacroExpander<'a, 'b> {
     monotonic: bool, // cf. `cx.monotonic_expander()`
 }
 
+pub fn expand_legacy_bang<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    key: (LocalExpnId, Span, LocalExpnId),
```

Comment:
The first `LocalExpnId` is unique for every macro call, so you won't get any reuse on this key.

Or you want to get reuse only between *different* compilations, when incremental is enabled?
In that case you do need the query system, not just a cache in resolver.

Reply:
I would be interested to see if there are any matching `ExpnId` hashes and any reuse at all in the simplest case (empty macro, two compilations, no changes between them).

Reply:
Yes, my main target is in two different compilations with incremental enabled. I think each macro should be expected to be expanded only once in a single compilation.

I cleaned up the `Span` in the key. Incremental compilation theoretically should not rely on any `Span`.

Yea, I think we need to do more testing of the effect of this change on two different compilations.